### PR TITLE
Learner luisa patch 1

### PR DESCRIPTION
**Pull request by luisas**
- Fix RayTuneLeaner class for tuning
- Add Class TuneTrain wrapper to handle the tuning and the training directly from the launcher in nf
- Add tests for all

There is still many functionalities missing - first prototype.

### CODE REVIEW

File: bin/launch_training.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+
+def main(config_path, model_path, data_path, experiment_path, output, best_config_path):
+    """
+    This launcher use ray tune to find the best hyperparameters for a given model.
+    """
+
+
+    # Create the learner
+    learner = StimulusTuneTrainWrapper(config_path, model_path, experiment_path, data_path)
```
Comment (alessiovignoli) [inline bin/launch_training.py:35]: why does it need the experiment?, what information is there needed for the tuning?

File: bin/src/data/csv.py [hunk 2]
```diff
@@ -198,10 +196,10 @@ def parse_csv_to_input_label_meta(self, load_method: Any) -> Tuple[dict, dict, d
             elif category.lower() == "label":
                 label_data[f"{name}:{data_type}"] = data[key]
             elif category.lower() == "meta":
-                meta_data[f"{name}:{data_type}"] = data[key]
+                meta_data[f"{name}"] = data[key]
```
Comment (alessiovignoli) [inline bin/src/data/csv.py:199]: why not data type on meta?, maybe just minor comment line above

File: bin/src/data/csv.py [hunk 3]
```diff
@@ -223,29 +222,48 @@ def get_and_encode(self, dictionary: dict, idx: Any) -> dict:
             # get the data at the given index
             # if the data is not a list, it is converted to a list
             # otherwise it breaks Float().encode_all(data) because it expects a list
-            data = dictionary[key][idx]
+            if idx is None:
```
Comment (alessiovignoli) [inline bin/src/data/csv.py:225]: please put data = Something explicitly above the if. I know python can recover lower level indentation variables in higher level indentation even if they were not initialized as such level. But that is just something python should'n't be doing to begin with.

File: bin/src/data/csv.py [hunk 4]
```diff
@@ -223,29 +222,48 @@ def get_and_encode(self, dictionary: dict, idx: Any) -> dict:
...
-    
+
+    def get_all_items(self) -> Tuple[dict, dict, dict]:
+        """
+        Returns all the items in the csv file, encoded.
+        TODO in the future we can optimize this for big datasets (ie. using batches, etc).
+        """
+        return self.get_and_encode(self.input), self.get_and_encode(self.label), self.meta
+
+    def get_all_items_and_length(self) -> Tuple[dict, dict, dict, int]:
```
Comment (alessiovignoli) [inline bin/src/data/csv.py:249]: why can not be there a different function to cumppute length? is there a reason id done here?

File: bin/src/data/handlertorch.py [hunk 5]
```diff
@@ -7,68 +7,63 @@
...
 class TorchDataset(Dataset):
     """
     Class for creating a torch dataset
     """
-    def __init__(self, csvpath : str, experiment : Any) -> None:
-        self.csvpath = csvpath
-        self.parser = CsvLoader(experiment, csvpath)
+    def __init__(self, csvpath: str, experiment: Any, split: Tuple[None, int] = None) -> None:
+        self.input, self.label, self.meta, self.length = CsvLoader(experiment, csvpath, split=split).get_all_items_and_length()
+        self.input, self.label = self.convert_dict_to_dict_of_tensors(self.input), self.convert_dict_to_dict_of_tensors(self.label)
```
Comment (alessiovignoli) [inline bin/src/data/handlertorch.py:18]: cmon put this on two different lines, is more readable.

File: bin/src/data/handlertorch.py [hunk 6]
```diff
@@ -7,68 +7,63 @@
...
-    def convert_list_of_numpy_arrays_to_tensor(self, data: list) -> Tuple[torch.Tensor, torch.Tensor]:
+    def convert_to_tensor(self, data: Union[np.ndarray, list], transform_method: Literal['pad_sequences'] = 'pad_sequences', **transform_kwargs) -> Union[torch.tensor, list]:
         """
-        Converts a list of numpy arrays to a tensor. 
-        If the list includes numpy arrays of different shapes, padd the numpy arrays and return a mask tensor, otherwise mask tensor is set to None
-
-        # TODO: This method utilizes ifs to check the shape of the data. this is not ideal. Performance improvement could be done here.
+        Converts the data to a tensor if the data is a numpy array.
+        Otherwise, when the data is a list, it calls a transform method to convert this list to a single pytorch tensor.
+        By default, this transformation method will padd 0 to the sequences to make them of the same length.
```
Comment (alessiovignoli) [inline bin/src/data/handlertorch.py:24]: is this what we want? which scenario is this possibility referring to?

File: bin/src/data/handlertorch.py [hunk 7]
```diff
@@ -7,68 +7,63 @@
...
     """
-    def __init__(self, csvpath : str, experiment : Any) -> None:
-        self.csvpath = csvpath
-        self.parser = CsvLoader(experiment, csvpath)
+    def __init__(self, csvpath: str, experiment: Any, split: Tuple[None, int] = None) -> None:
+        self.input, self.label, self.meta, self.length = CsvLoader(experiment, csvpath, split=split).get_all_items_and_length()
+        self.input, self.label = self.convert_dict_to_dict_of_tensors(self.input), self.convert_dict_to_dict_of_tensors(self.label)
 
-    def convert_list_of_numpy_arrays_to_tensor(self, data: list) -> Tuple[torch.Tensor, torch.Tensor]:
+    def convert_to_tensor(self, data: Union[np.ndarray, list], transform_method: Literal['pad_sequences'] = 'pad_sequences', **transform_kwargs) -> Union[torch.tensor, list]:
```
Comment (alessiovignoli) [inline bin/src/data/handlertorch.py:20]: transform method is referring to padding. In the future are going to be there other methods? if so they will be handled all by the Literal[keywords] definition?

File: bin/src/learner/raytune_learner.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        Prepare the tuner with the configs.
+        """
+        self.tuner = tune.Tuner(TuneModel,
+                            tune_config= self.tune_config,
+                            param_space=self.config,
+                            run_config=self.run_config,
+                        )
+        self.tuner.tuning = True
+
+    def tune(self, overwrite=False):
```
Comment (alessiovignoli) [inline bin/src/learner/raytune_learner.py:78]: Is this an override of the imported tune function from torch? if not is better is not spelled identically, i know is nitpicky but better to be safe.

File: bin/src/learner/raytune_learner.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+        """
+        Run the tuning process.
+        """
+        if overwrite:
```
Comment (alessiovignoli) [inline bin/src/learner/raytune_learner.py:82]: describe why this is here

File: bin/src/learner/raytune_learner.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        module_name, class_name = config["model"].rsplit('.', 1)
+        module = importlib.import_module(module_name)
+        model = getattr(module, class_name)
+        self.model = model(**config["model_params"])
+
+        # Add data path
+        self.data_path = config["data_path"]
+
+        # Add experiment
+        module_name, class_name = config["experiment"].rsplit('.', 1)
```
Comment (alessiovignoli) [inline bin/src/learner/raytune_learner.py:140]: why is overwriting the line 131?

Comment (alessiovignoli) [reply bin/src/learner/raytune_learner.py:140]: or better why is importing from the experiment keyword as well?

File: bin/src/learner/raytune_learner.py [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+        self.loss_dict = config["loss_fn"]
+        for key, loss_fn in self.loss_dict.items():
+            try:
+                #self.loss_dict[key] = getattr(nn, loss_fn)()
+                self.loss_dict[key] = loss_fn
+            except AttributeError:
+                raise ValueError(f"Invalid loss function: {loss_fn}, check PyTorch for documentation on available loss functions")
+        
+        # get the optimizer from the config
+        try:
```
Comment (alessiovignoli) [inline bin/src/learner/raytune_learner.py:156]: can this two checks, this line and line 149 be moved in the TuneWrapper like all the other checks?

File: bin/launch_training.py (hunk 1 above)
Comment (alessiovignoli) [reply bin/launch_training.py:35]: found out. Tune calls -> TorchDataset  that calls -> CsvLoader that calls -> experiment to know how the user wants to encode each column.

File: bin/src/learner/raytune_learner.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+
+        # get learning rate
+        self.lr = config['lr']
+
+        # get the train and validation data from the config
+        # run dataloader on them
+        self.training = DataLoader(TorchDataset(self.data_path, self.experiment, split=0), batch_size=config['data_params']['batch_size'])
+        self.validation = DataLoader(TorchDataset(self.data_path, self.experiment, split=1), batch_size=config['data_params']['batch_size'])
+
+        self.tuning = False
```
Comment (alessiovignoli) [inline bin/src/learner/raytune_learner.py:175]: comment line

File: bin/tests/test_raytune_learner.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        checkpoint = self.learner.save_checkpoint("bin/tests/test_data/dna_experiment/test_checkpoint.pth")
+        self.assertIsInstance(checkpoint, dict)
+        self.assertTrue(os.path.exists("bin/tests/test_data/dna_experiment/test_checkpoint.pth"))
+        os.remove("bin/tests/test_data/dna_experiment/test_checkpoint.pth")
+
+    def test_load_checkpoint(self):
+        self.learner.save_checkpoint("bin/tests/test_data/dna_experiment/test_checkpoint.pth")
+        checkpoint = self.learner.save_checkpoint("bin/tests/test_data/dna_experiment/test_checkpoint.pth")
+        self.learner.load_checkpoint(checkpoint)
+        self.assertTrue(True)
```
Comment (alessiovignoli) [inline bin/tests/test_raytune_learner.py:62]: insightfull

**Review by alessiovignoli — CHANGES_REQUESTED** [review]
check comments

File: bin/src/data/csv.py (hunk 2 above)
Comment (mathysgrapotte) [reply bin/src/data/csv.py:199]: The type of meta should not matter for the training, however, we have to think of a way to make it work for downstream analysis.

File: bin/src/learner/raytune_learner.py (hunk 8 above)
Comment (luisas) [reply bin/src/learner/raytune_learner.py:78]: No no, it is the une function of the tune wrapper - so no overwriting :)

File: bin/src/data/csv.py (hunk 2 above)
Comment (suzannejin) [reply bin/src/data/csv.py:199]: Since this is the csvloader, it should be general, so meta should still keep data type

File: bin/src/learner/raytune_learner.py (hunk 9 above)
Comment (luisas) [reply bin/src/learner/raytune_learner.py:82]: Ok done :)

File: bin/src/learner/raytune_learner.py (hunk 10 above)
Comment (luisas) [reply bin/src/learner/raytune_learner.py:140]: because it is imported as a string first and then we need to import the actual class

File: bin/src/learner/raytune_learner.py (hunk 11 above)
Comment (luisas) [reply bin/src/learner/raytune_learner.py:156]: Yes i think it could - yet i'd leave this whole moving of the checking to whoever does it for the corresponding issue so it's done once and for all. But good point!

File: bin/tests/test_raytune_learner.py (hunk 13 above)
Comment (luisas) [reply bin/tests/test_raytune_learner.py:62]: ahahah fair point - deleted

File: bin/src/data/csv.py (hunk 3 above)
Comment (suzannejin) [reply bin/src/data/csv.py:225]: why?

File: bin/src/data/handlertorch.py (hunk 7 above)
Comment (mathysgrapotte) [reply bin/src/data/handlertorch.py:20]: Yes

**Review by alessiovignoli — APPROVED** [review]
